pkg/skaffold/initializer/init.go

```diff
@@ -538,40 +542,80 @@ func printAnalyzeJSON(out io.Writer, skipBuild bool, pairs []builderImagePair, u
 	return err
 }
 
-func walk(dir string, force, enableJibInit bool, validateBuildFile func(bool, string) ([]InitBuilder, error)) ([]string, []InitBuilder, error) {
+func walk(dir string, force, enableJibInit bool) ([]string, []InitBuilder, error) {
 	var potentialConfigs []string
 	var foundBuilders []InitBuilder
-	err := filepath.Walk(dir, func(path string, f os.FileInfo, e error) error {
-		if f.IsDir() && util.IsHiddenDir(f.Name()) {
-			logrus.Debugf("skip walking hidden dir %s", f.Name())
-			return filepath.SkipDir
-		}
-		if f.IsDir() || util.IsHiddenFile(f.Name()) {
-			return nil
+
+	var enterDirectory func(path string, findBuilders bool) error
+	enterDirectory = func(path string, findBuilders bool) error {
+		dirents, err := godirwalk.ReadDirents(path, nil)
+		if err != nil {
+			return err
 		}
-		if IsSkaffoldConfig(path) {
-			if !force {
-				return fmt.Errorf("pre-existing %s found", path)
+
+		var directories []*godirwalk.Dirent
+		findBuildersInDirectories := true
+		sort.Sort(dirents)
+
+		// Traverse files
+		for _, file := range dirents {
+			if util.IsHiddenFile(file.Name()) || util.IsHiddenDir(file.Name()) {
+				continue
+			}
+
+			// If we found a directory, keep track of it until we've gone through all the files first
+			if file.IsDir() {
+				directories = append(directories, file)
+				continue
+			}
+
+			filePath := filepath.Join(path, file.Name())
+			if findBuildersInDirectories, err = checkFile(filePath, force, findBuilders, enableJibInit, &potentialConfigs, &foundBuilders); err != nil {
```

Comment:
I think `findBuildersInDirectories` should be some sort of accumulation that assignment. It can't be that you just overwrite the previous value with the most recent, no?

Reply:
In practice that will never happen in an unintended way, but I can try to make the code more clear about it.

Reply:
Hmm... I think the problem I found is like this. Let's say we have two files in the root directory:

`/fileA`: `detectBuilders` inside `checkFile` will return true for `findBuildersInSubdirectories`.
`/fileB`: `detectBuilders` will return false.

Let's assume `findBuilders` is true at the root loop. Note this `findBuilders` doesn't change inside the loop, always being true.

So, if we check the files in the order A -> B in the loop, `findBuildersInSubdirectories` will end up being false after the loop terminates. OTOH, if the loop goes B -> A, it will end up being true, because the last `checkFile` assignment will win.

And this was the problem that I had originally pointed out. Then I told you later offline that this seemed okay, but looking at this again, now I think this problem still stands for both before and after your last commit.

Reply:
I think I finally see what you mean by accumulation now, and I think you're right. Ugh. I will fix it soon.